### PR TITLE
unit test coverage for PubSubClient NATS implementation Publish method

### PR DESCRIPTION
**Pull request by aaslamin**
#### Parent issue: https://github.com/aporeto-inc/aporeto/issues/1274

This PR provides close to 💯% unit test coverage on the [`Publish` method](https://github.com/aporeto-inc/bahamut/blob/master/pubsub_nats.go#L58) for the NATS implementation of the [`PubSubClient` interface](https://github.com/aporeto-inc/bahamut/blob/master/pubsub.go#L25).

Coming up in future PRs:

- Unit tests for the NATS `Subscribe` method
- Unit tests for NATS `Connect` method
- Unit tests for NATS `Disconnect` method

### CODE REVIEW

File: pubsub_nats_test.go [hunk 1]
```diff
@@ -30,7 +41,12 @@ func TestNats_NewPubSubServer(t *testing.T) {
 			So(ps.clientID, ShouldNotBeEmpty)
 			So(ps.username, ShouldEqual, "")
 			So(ps.password, ShouldEqual, "")
+			So(ps.retryInterval, ShouldBeGreaterThan, 0)
 			So(ps.tlsConfig, ShouldEqual, nil)
+			// verify that client id is a proper V4 UUID
+			id, err := uuid.FromString(ps.clientID)
+			So(err, ShouldBeNil)
+			So(id.Version(), ShouldEqual, uuid.V4)
```
Comment (aaslamin) [inline pubsub_nats_test.go]: 👀 this will prevent someone in the future changing behaviour by not using a UUID for the client ID (that is autogenerated).

**Comment by aaslamin** [comment]
Cover all teh things!

**Comment by aaslamin** [comment]
This fix needs to go in first: https://github.com/aporeto-inc/elemental/pull/42

File: pubsub_nats.go [hunk 2]
```diff
@@ -16,16 +16,35 @@ import (
...
+}
+
+func (e *EncodingError) Error() string {
+	return fmt.Sprintf("%s: %s", e.Message, e.Err)
+}
+
 type natsPubSub struct {
 	natsURL        string
-	client         *nats.Conn
+	client         NATSClient
```
Comment (primalmotion) [inline pubsub_nats.go]: I don't think this interface needs to be public

File: pubsub_nats_options.go [hunk 3]
```diff
@@ -52,6 +60,14 @@ func NATSOptTLS(tlsConfig *tls.Config) NATSOption {
 	}
 }
 
+// NATSOptClient sets the NATS client that will be used
+// This is useful for unit testing as you can pass in a mocked NATS client
+func NATSOptClient(client NATSClient) NATSOption {
```
Comment (primalmotion) [inline pubsub_nats_options.go]: this should also be private

File: pubsub_nats_test.go [hunk 4]
```diff
@@ -12,13 +12,24 @@
...
 import (
+	"context"
 	"crypto/tls"
+	"errors"
+	"reflect"
 	"testing"
+	"time"
 
+	"go.aporeto.io/elemental"
+
```
Comment (primalmotion) [inline pubsub_nats_test.go]: no new lines here

File: pubsub_nats_options.go (hunk 3 above)
Comment (aaslamin) [reply pubsub_nats_options.go]: Done

File: pubsub_nats_test.go (hunk 4 above)
Comment (aaslamin) [reply pubsub_nats_test.go]: Done

File: pubsub_nats.go (hunk 2 above)
Comment (aaslamin) [reply pubsub_nats.go]: Done

**Comment by aaslamin** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "nats-pubsub-publish-tests",
    "component": "bahamut",
    "pr-id": "52",
    "commit-sha": "d46a71ad26649bd32930ba4f322bf08e798288ed"
  }
]
```

**Comment by aaslamin** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "nats-pubsub-publish-tests",
    "component": "bahamut",
    "pr-id": "52",
    "commit-sha": "aa115467b05afff8a0b14e9be152e1224f0b9509"
  }
]
```

**Pull request merged**